pypeit/reduce.py

```diff
@@ -389,9 +389,20 @@ def run(self, ra=None, dec=None, obstime=None, std_trace=None, show_peaks=False,
             msgs.info("Skipping 2nd run of finding objects")
 
         # Assign here -- in case we make another pass to add in missing targets
-        if self.nobj > 0 and self.par['reduce']['slitmask']['assign_obj'] and self.slits.maskdef_designtab is not None:
-            self.slits.assign_maskinfo(self.sobjs_obj, self.get_platescale(None),
-                                       TOLER=self.par['reduce']['slitmask']['obj_toler'])
+        if self.par['reduce']['slitmask']['assign_obj'] and self.slits.maskdef_designtab is not None:
+            msgs.info('Assign slitmask design info to detected objects')
+            all_expected_objpos = self.slits.assign_maskinfo(self.sobjs_obj, self.get_platescale(None),
+                                                             self.slits_left, self.slits_right,
+                                                             self.par['calibrations']['slitedges']['det_buffer'],
+                                                             TOLER=self.par['reduce']['slitmask']['obj_toler'])
+            # force extraction of non detected objects
+            if self.par['reduce']['slitmask']['force_extract']:
```

Comment:
I am not following why you are not using the manual object infrastructure here, and instead have created a separate framework for adding objects. Is there any difference between what you are doing here and adding manual objects? I agree that the current framework is not very clean or easy to understand, but we basically have two ways of manually adding objects to slits, which is not very desirable.

Reply:
I disagree.  The `manual` approach is qualitatively different than assigning by mask design.

It would be useful, however, to have a single method in `SpecObj` for vetting the addition of a new object (took me awhile to get all the required bits and pieces).